### ISSUE TITLE
[discussion] Add `f16`-friendly operations

### ISSUE THREAD

**Issue by swfsql** [tracel-ai/burn#4235]
### Feature description

Implement alternative branches specialized on the lower dynamic range of `f16` for cases when that dtype doesn't get cast/mixed to others such as `bf16` and `f32`. 

### Feature motivation

It's often hard to accommodate operations in `f16` since it has a lower dynamic/exponential range. E.g. values seem as "normal" can saturate or overflow, such as `256 * 256` and `e ^ 12`.  
Assuming alternative branches for `f16` are available, it would still be hard to decide which particular operations should adopt those alternative branches, and to what level they should do so. Even operations that get adapted to a lower range still have their own range limits, and they may have shifted their limits from "overflow" to a "loss of precision", and they most likely have additional steps in calculation.  
Given that this is hard to know for specific operations, then there isn't a default way to decide whether a `f16` operation should use the default `f32`-like sequence of operations, or the specialized ones. The same kind of operation may have different exponential range vs precision requirements depending on their inputs (their position on the code) and the current weight values (which vary over the training steps), so this is likely not something that is solved statically. So a good design to get the best of both words is unknown to me.

This being said, this could be useful if the model design insists in not casting nor mixing to alternative dtypes that has a higher dynamic range such as `bf16` and `f32`. It may as well be the case that casting/mixing is both simpler and preferential to adapting operations to strictly stay in `f16`.

### (Optional) Suggest a Solution

One option is to do a dtype-based branching -- the code doesn't look particularly good, but it's a start :)

Examples:

<details> 
  <summary>softplus_no_beta</summary>

```rust
pub fn softplus_no_beta<const D: usize, B: Backend>(x: Tensor<B, D>) -> Tensor<B, D> {
    match <B::FloatElem as Element>::dtype() {
        DType::F64 | DType::F32 | DType::Flex32 | DType::BF16 => {
            // softplus = log(e^x + 1)
            x.exp().log1p()
        }
        DType::F16 => {
            // max(a,b) = (a + b + |a - b|) / 2
            // softplus = max(x, 0) + log(e^-|x| + 1) 
            //          = (x + |x|) / 2 + log(e^-|x| + 1)
            //
            // this avoids e^x, which →INF when x ≃ 11,
            // since e ^ -|x| <= 1
            let xabs = x.clone().abs();
            (x + xabs.clone()) / 2. + xabs.neg().exp().log1p()
        }
        DType::I64 | DType::I32 | DType::I16 | DType::I8 | DType::U64 | DType::U32 | DType::U16 | DType::U8 | DType::Bool => {
            unreachable!()
        }
        DType::QFloat(_) => {
            unimplemented!()
        }
    }
}
```

</details>

<details> 
  <summary>rms_norm_forward</summary>

```rust
pub fn rms_norm_forward<const D: usize>(&self, x: Tensor<B, D>) -> Tensor<B, D> {
    match <B::FloatElem as Element>::dtype() {
        DType::F64 | DType::F32 | DType::Flex32 | DType::BF16 => {
            let div_eps = div_eps::<B>();
            let rms =
                (x.clone() * x.clone()).mean_dim(D - 1).sqrt();
            let normalized = (x / (rms + div_eps)) * self.gamma.val().unsqueeze();
            normalized
        }
        DType::F16 => {
            use burn::tensor::ElementConversion; 
            let div_eps: half::f16 = half::f16::from_elem(div_eps::<B>()) * half::f16::from_f32(2.);
            // avoid calculating x² directly (due to overflow e.g. on 256 * 256)
            let max = x.clone().no_grad().detach().abs().max().expand(x.shape());
            let x_ = x.clone() / (max.clone() + div_eps); // x_.abs() <= 1
            let rms_partial = (x.clone() * x_).mean_dim(D - 1).sqrt(); // √(x²) = √(x²/max) * √max
            let normalized = (x / (rms_partial + div_eps)) / max.sqrt() * self.gamma.val().unsqueeze();
            normalized
        }
        DType::I64 | DType::I32 | DType::I16 | DType::I8 | DType::U64 | DType::U32 | DType::U16 | DType::U8 |  => DType::Bool {
            unreachable!()
        }
        DType::QFloat(_) => {
            unimplemented!()
        }
    }
}
```

</details>

<details> 
  <summary>div_eps (suggestion)</summary>

```rust
pub fn div_eps<B: Backend>() -> B::FloatElem {
    match <B::FloatElem as Element>::dtype() {
        // 4.0693917e-16
        DType::F64 => {
            let raw_exp = -(-f64::MIN_EXP as f32 * 2.3f32).powf(0.35f32);
            let eps_exp = (f64::EPSILON as f32).log10();
            let avg = (raw_exp + eps_exp) / 2f32;
            10f32.powf(avg).elem()
        }
        // 8.1584695e-8
        DType::F32 | DType::Flex32 => {
            let raw_exp = -(-f32::MIN_EXP as f32 * 2.3f32).powf(0.35f32);
            let eps_exp = f32::EPSILON.log10();
            let avg = (raw_exp + eps_exp) / 2f32;
            10f32.powf(avg).elem()
        }
        // 7.1209995e-4
        DType::F16 => {
            let raw_exp = -(-burn::tensor::f16::MIN_EXP.to_f32() * 2.3f32).powf(0.35f32);
            let eps_exp = burn::tensor::f16::EPSILON.to_f32().log10();
            let avg = (raw_exp + eps_exp) / 2f32;
            10f32.powf(avg).elem()
        }
        // 2.0885676e-5;
        DType::BF16 => {
            let raw_exp = -(-burn::tensor::bf16::MIN_EXP.to_f32() * 2.3f32).powf(0.35f32);
            let eps_exp = burn::tensor::bf16::EPSILON.to_f32().log10();
            let avg = (raw_exp + eps_exp) / 2f32;
            10f32.powf(avg).elem()
        }
        DType::I64 | DType::I32 | DType::I16 | DType::I8 | DType::U64 | DType::U32 | DType::U16 | DType::U8 => {
            unreachable!()
        }
        DType::Bool => {
            unreachable!()
        }
        DType::QFloat(_) => {
            unimplemented!()
        }
    }
}
```

</details>

Alternatively, traits could represent those different dtypes and/or range-vs-precision requirements, and allow tensor operations to implement specialized methods that could fallback to a default (f32-like) implementation.

**Comment by swfsql** [comment]
Accumulators in f16 are being done in f32, so I think this largely solves this issue. F16-specific cases may still exist out of accumulators, but I think this just fallbacks to case-by-case custom implementations. Thanks!
